lib/ecto/query/builder/from.ex

```diff
@@ -50,6 +50,9 @@ defmodule Ecto.Query.Builder.From do
   end
 
   defp escape_source(query, _env), do: query
+  
+  @typep dynamic_hint :: {atom, term}
+  @typep hints :: String.t() | dynamic_hint | [String.t() | dynamic_hint]
```

Comment:
Hints is always a list or nil, so we want this:

```suggestion
  @typep hints :: [String.t() | {atom, term}]
```